src/sysevent.c

```diff
@@ -0,0 +1,984 @@
+      sysevent_thread_error = 1;
+      break;
+    }
+
+    if (sysevent_thread_loop <= 0)
+      break;
+  } /* while (sysevent_thread_loop > 0) */
+
+  pthread_mutex_unlock(&sysevent_lock);
+
+  // pthread_exit instead of return?
+  return ((void *)0);
+} /* }}} void *sysevent_thread */
+
+static int start_thread(void) /* {{{ */
+{
+  int status;
+
+  pthread_mutex_lock(&sysevent_lock);
+
+  if (sysevent_thread_loop != 0) {
+    pthread_mutex_unlock(&sysevent_lock);
+    return (0);
+  }
+
+  sysevent_thread_loop = 1;
+  sysevent_thread_error = 0;
+
+  DEBUG("sysevent plugin: starting thread");
+
+  status = plugin_thread_create(&sysevent_thread_id, /* attr = */ NULL,
+                                sysevent_thread,
+                                /* arg = */ (void *)0, "sysevent");
+  if (status != 0) {
+    sysevent_thread_loop = 0;
+    ERROR("sysevent plugin: starting thread failed.");
+    pthread_mutex_unlock(&sysevent_lock);
+    return (-1);
+  }
+
+  pthread_mutex_unlock(&sysevent_lock);
+  return (0);
+} /* }}} int start_thread */
+
+static int stop_thread(int shutdown) /* {{{ */
+{
+  int status;
+
+  pthread_mutex_lock(&sysevent_lock);
+
+  if (sysevent_thread_loop == 0) {
+    pthread_mutex_unlock(&sysevent_lock);
+    return (-1);
+  }
+
+  sysevent_thread_loop = 0;
+  pthread_mutex_unlock(&sysevent_lock);
+
+  if (shutdown == 1) {
+    // Since the thread is blocking, calling pthread_join
+    // doesn't actually succeed in stopping it.  It will stick around
+    // until a message is received on the socket (at which
+    // it will realize that "sysevent_thread_loop" is 0 and will
+    // break out of the read loop and be allowed to die).  This is
+    // fine when the process isn't supposed to be exiting, but in
+    // the case of a process shutdown, we don't want to have an
+    // idle thread hanging around.  Calling pthread_cancel here in
+    // the case of a shutdown is just assures that the thread is
+    // gone and that the process has been fully terminated.
+
+    DEBUG("sysevent plugin: Canceling thread for process shutdown");
+
+    status = pthread_cancel(sysevent_thread_id);
+
+    if (status != 0) {
+      ERROR("sysevent plugin: Unable to cancel thread: %d (%s)", status,
+            strerror(errno));
+      status = -1;
+    }
+  } else {
+    status = pthread_join(sysevent_thread_id, /* return = */ NULL);
+    if (status != 0) {
+      ERROR("sysevent plugin: Stopping thread failed.");
+      status = -1;
+    }
+  }
+
+  pthread_mutex_lock(&sysevent_lock);
+  memset(&sysevent_thread_id, 0, sizeof(sysevent_thread_id));
+  sysevent_thread_error = 0;
+  pthread_mutex_unlock(&sysevent_lock);
+
+  DEBUG("sysevent plugin: Finished requesting stop of thread");
+
+  return (status);
+} /* }}} int stop_thread */
+
+static int sysevent_init(void) /* {{{ */
+{
+  ring.head = 0;
+  ring.tail = 0;
+  ring.maxLen = buffer_length;
+  ring.buffer = (char **)malloc(buffer_length * sizeof(char *));
```

Comment:
return value of malloc should be checked if NULL

Reply:
Thanks again for spotting this one too.